### PR TITLE
ns: rewrite for new IPC

### PR DESCRIPTION
**Pull request by liamwhite**
Rewrites and adds new functionality for partial qlaunch support

### CODE REVIEW

File: src/core/hle/service/ns/ns_types.h [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    ApplicationRecordType type;
+    u8 unknown;
+    INSERT_PADDING_BYTES_NOINIT(0x6);
+    u8 unknown2;
+    INSERT_PADDING_BYTES_NOINIT(0x7);
+};
+static_assert(sizeof(ApplicationRecord) == 0x18, "ApplicationRecord is an invalid size");
+
+/// ApplicationView
+struct ApplicationView {
```
Comment (FearlessTobi) [inline src/core/hle/service/ns/ns_types.h:43]: missing a size assert. also we use std:arrays

File: src/core/hle/service/ns/ns_types.h [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    u16 unk_x24;        ///< Unknown.
+    u8 unk_x26[0x2];    ///< Unknown.
+    u8 unk_x28[0x8];    ///< Unknown.
+    u8 unk_x30[0x10];   ///< Unknown.
+    u32 unk_x40;        ///< Unknown.
+    u8 unk_x44;         ///< Unknown.
+    u8 unk_x45[0xb];    ///< Unknown.
+};
+
+struct ApplicationRightsOnClient {
```
Comment (FearlessTobi) [inline src/core/hle/service/ns/ns_types.h:58]: dito

File: src/core/hle/service/ns/ns_types.h [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    u64 application_id;
+    Common::UUID uid;
+    u8 flags;
+    u8 flags2;
+    INSERT_PADDING_BYTES(0x6);
+};
+
+/// NsPromotionInfo
+struct PromotionInfo {
```
Comment (FearlessTobi) [inline src/core/hle/service/ns/ns_types.h:67]: same

File: src/core/hle/service/ns/ns_types.h [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    s64 remaining_time;  ///< Remaining time until the promotion ends, in nanoseconds
+                         ///< ({end_timestamp - current_time} converted to nanoseconds).
+    INSERT_PADDING_BYTES_NOINIT(0x4);
+    u8 flags; ///< Flags. Bit0: whether the PromotionInfo is valid (including bit1). Bit1 clear:
+              ///< remaining_time is set.
+    INSERT_PADDING_BYTES_NOINIT(0x3);
+};
+
+/// NsApplicationViewWithPromotionInfo
+struct ApplicationViewWithPromotionInfo {
```
Comment (FearlessTobi) [inline src/core/hle/service/ns/ns_types.h:79]: same

File: src/core/hle/service/ns/ns_types.h [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+struct ApplicationOccupiedSizeEntity {
+    FileSys::StorageId storage_id;
+    u64 app_size;
+    u64 patch_size;
+    u64 aoc_size;
+};
+static_assert(sizeof(ApplicationOccupiedSizeEntity) == 0x20,
+              "ApplicationOccupiedSizeEntity has incorrect size.");
+
+struct ApplicationOccupiedSize {
```
Comment (FearlessTobi) [inline src/core/hle/service/ns/ns_types.h:93]: same

File: src/core/hle/service/ns/ns_types.h [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    std::array<ApplicationOccupiedSizeEntity, 4> entities;
+};
+
+struct ContentPath {
```
Comment (FearlessTobi) [inline src/core/hle/service/ns/ns_types.h:97]: same

File: src/core/hle/service/ns/read_only_application_control_data_interface.cpp [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    LOG_INFO(Service_NS, "called with control_source={}, application_id={:016X}",
+             application_control_source, application_id);
+
+    const FileSys::PatchManager pm{application_id, system.GetFileSystemController(),
+                                   system.GetContentProvider()};
+    const auto control = pm.GetControlMetadata();
+    const auto size = out_buffer.size();
+
+    const auto icon_size = control.second ? control.second->GetSize() : 0;
+    const auto total_size = 0x4000 + icon_size;
```
Comment (FearlessTobi) [inline src/core/hle/service/ns/read_only_application_control_data_interface.cpp:46]: do we know what this magic number is? if yes then make it a named variable. same below

**Review by FearlessTobi — COMMENTED** [review]
Took a quick look and didn't see any glaring issues. Just some nits:

**Review by german77 — APPROVED** [review]

**Pull request merged**